pkg/proxy/ipvs/netlink_linux.go

```diff
@@ -123,72 +124,67 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
-// 10.0.0.252 dev kube-ipvs0  scope host  src 10.0.0.252
-// 100.106.89.164 dev eth0  scope host  src 100.106.89.164
-// 127.0.0.0/8 dev lo  scope host  src 127.0.0.1
-// 127.0.0.1 dev lo  scope host  src 127.0.0.1
-// 172.17.0.1 dev docker0  scope host  src 172.17.0.1
-// 192.168.122.1 dev virbr0  scope host  src 192.168.122.1
-// Then cut the unique src IP fields,
-// --> result set: [10.0.0.1, 10.0.0.10, 10.0.0.252, 100.106.89.164, 127.0.0.1, 172.17.0.1, 192.168.122.1]
-
-// If dev is specified, it's equivalent to exec:
-// $ ip route show table local type local proto kernel dev kube-ipvs0
-// 10.0.0.1  scope host  src 10.0.0.1
-// 10.0.0.10  scope host  src 10.0.0.10
-// Then cut the unique src IP fields,
-// --> result set: [10.0.0.1, 10.0.0.10]
-
-// If filterDev is specified, the result will discard route of specified device and cut src from other routes.
+// GetLocalAddresses return addresses for interfaces in two ways;
+//  1. dev != "" && filterDev == ""; Return addresses for `dev`
+//  2. dev == "" && filterDev != ""; Return addresses for all interfaces *except* filterDev
+// Other combinations of dev and filterDev are forbidden.
+// IPv6 link-local addresses are excluded.
 func (h *netlinkHandle) GetLocalAddresses(dev, filterDev string) (sets.String, error) {
-	chosenLinkIndex, filterLinkIndex := -1, -1
 	if dev != "" {
-		link, err := h.LinkByName(dev)
-		if err != nil {
-			return nil, fmt.Errorf("error get device %s, err: %v", dev, err)
+		if filterDev != "" {
+			return nil, fmt.Errorf("Invalid argument")
 		}
-		chosenLinkIndex = link.Attrs().Index
-	} else if filterDev != "" {
-		link, err := h.LinkByName(filterDev)
+		ifi, err := net.InterfaceByName(dev)
 		if err != nil {
-			return nil, fmt.Errorf("error get filter device %s, err: %v", filterDev, err)
+			return nil, fmt.Errorf("Could not get interface %s: %v", dev, err)
 		}
-		filterLinkIndex = link.Attrs().Index
-	}
-
-	routeFilter := &netlink.Route{
-		Table:    unix.RT_TABLE_LOCAL,
-		Type:     unix.RTN_LOCAL,
-		Protocol: unix.RTPROT_KERNEL,
+		ips := sets.NewString()
+		h.appendAddresses(ips, ifi)
+		return ips, nil
 	}
-	filterMask := netlink.RT_FILTER_TABLE | netlink.RT_FILTER_TYPE | netlink.RT_FILTER_PROTOCOL
 
-	// find chosen device
-	if chosenLinkIndex != -1 {
-		routeFilter.LinkIndex = chosenLinkIndex
-		filterMask |= netlink.RT_FILTER_OIF
+	if filterDev == "" {
+		return nil, fmt.Errorf("Invalid argument")
 	}
-	routes, err := h.RouteListFiltered(netlink.FAMILY_ALL, routeFilter, filterMask)
+	ifaces, err := net.Interfaces()
 	if err != nil {
-		return nil, fmt.Errorf("error list route table, err: %v", err)
+		return nil, fmt.Errorf("Could not get interfaces: %v", err)
 	}
-	res := sets.NewString()
-	for _, route := range routes {
-		if route.LinkIndex == filterLinkIndex {
+	ips := sets.NewString()
+	for _, i := range ifaces {
+		if i.Name == filterDev {
+			continue
+		}
+		h.appendAddresses(ips, &i)
+	}
+	return ips, nil
+}
+
```

Comment:
a personal thought, is not clearer to have a function
```
func getInterfaceAddresses(ifi *net.Interface) []string {
}
```
and do the set insertions in the main loop?

Reply:
Good point. I'll check that :+1:

Reply:
Would remove the reference-issue also.

Reply:
checking in, do you prefer current format and no try to return a slice instead of  set?